Adapt content to outbreak context and R packages (#13)
epiverse-trace/git-rstudio-basics: episodes/02-setup.md

---
title: Setting Up Git
teaching: 10
exercises: 0
---

::::::::::::::::::::::::::::::::::::::: objectives

- Configure `git` the first time it is used on a computer.

::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: questions

> - How do I get set up to use Git?

Reviewer: Would "How do I set up Git" be a simpler phrasing?

Author: Edited in https://github.com/epiverse-trace/git-rstudio-basics/pull/13/commits/cb3f384b89dc6fe9285f1a4de9434057b76216f3

> - How do I get set up to use Git?

Author: ```suggestion
- How do I set up Git?
```

::::::::::::::::::::::::::::::::::::::::::::::::::

-->

## Set up your GitHub token

To interact with GitHub we need to include credentials in the request. We are going to configure one type of credential called _Personal Access Token (PAT)_. We need this to prove that we are a specific GitHub user, allowed to do whatever we’re asking to do. ([Bryan, 2021](https://happygitwithr.com/https-pat.html))

__First, let's create your token.__

Do this with `usethis::create_github_token()`. This function should redirect you to GitHub on your browser. Once there, check all the options in the figure below.

```r
usethis::create_github_token()
```

Describe the token use case.

> Check all of the following options:

Author: ```suggestion
Look over the scopes; the recommended scopes to select are “repo”, “user”, and “workflow”. When using `usethis::create_github_token()` will pre-select all the recommended scopes for you!

 Click “Generate token”.
```

> ![](fig/git-token.png)

Reviewer: It might be worth explaining some of the details of the options they are selecting. Nothing too detailed, but just to give a very short overview of what this PAT will allow them to do.

Author: I agree. I added text in https://github.com/epiverse-trace/git-rstudio-basics/pull/13/commits/4ad447615fd9a70658f12f29d366a05c0c7699ee and a callout https://github.com/epiverse-trace/git-rstudio-basics/pull/13/commits/0cb5ebe621fdc23fa16d45f0d19e920da48084ab

> ![](fig/git-token.png)

Author: ```suggestion
![Visual display with the recommended scopes selected.](fig/git-token.png)
```

1: Abort update with error, and keep the existing credentials
2: Replace these credentials
3: See the password / token

Selection: 2

```

Paste your `token` to save it and complete this step.

__Lastly, let's confirm your setting.__

Run:

```r
usethis::git_sitrep()
```

In the `── Git global (user)` section, the two first lines of the output should look like this:

```output
── Git global (user) 
• Name: 'Vlad Dracula'
• Email: 'vlad@tran.sylvan.ia'
```

In the `── GitHub user` section, the three first lines of the output should look like this:

```output
── GitHub user 
• Default GitHub host: 'https://github.com'
• Personal access token for 'https://github.com': '<discovered>'
• GitHub user: 'vlad'
```

<!--
```r
gh::gh_whoami()
```

```output
{
  "name": "Vlad Dracula",
  "login": "vlad",
  "html_url": "https://github.com/vlad",
  "scopes": "gist, repo, workflow",
  "token": "gho_...AlAn"
} 
```
-->

You should recognize your:

- Name,
- GitHub login, and
- Token.

## Set up a default `branch` name

As we mentioned before, the complete history of `commits` for a particular project and their metadata make up a `repository`. (We are going to create one in the next episode!)

> A [`branch`](https://glosario.carpentries.org/en/#git_branch) is a snapshot of a version of a repository. In that sense, a repository can have more that one branch. WHAT?!! How is that possible? We are going to see that in coming episodes!

Reviewer: ```suggestion
A [`branch`](https://glosario.carpentries.org/en/#git_branch) is a snapshot of a version of a repository. In that sense, a repository can have more than one branch. WHAT?!! How is that possible? We are going to see that in coming episodes!
```

$ git config --list
```

::::::::::::::::::::::::::::::
-->

And if necessary, change your configuration using the
same commands to update your email address.
This can be done as many times as you want.

:::::::::::::::::::::::::::::::::::::::::  callout

> ## Proxy

Reviewer: This section feels isolated. I have never used a proxy for Git, so I do not fully understand when one would be necessary. Is this information critical for new Git users? If not I would remove it. If it is crucial information I would explain when this step would be required in more detail.

Author: I agree, I opted remove it's visibility https://github.com/epiverse-trace/git-rstudio-basics/pull/13/commits/f7bf947f6c2967e774c1a983e26f9a78288e9925

::::::::::::::::::::::::::::::::::::::::::::::::::

-->

::::::::::::::::: instructor

When using `usethis::git_sitrep()`, check if there is no `✖ ...` line in the output with an error message. 

Some common errors are solved here:

- [Happy Git with R ebook](https://happygitwithr.com/https-pat.html#pat-troubleshooting)
- [Workbench documentation](https://carpentries.github.io/sandpaper-docs/github-pat.html#common-token-errors)

:::::::::::::::::::::::::

::::::::::::::::: callout

> When using the [Terminal](https://glosario.carpentries.org/en/#console), this steps is known as `git config` for `--global` option. In the next chapter we are going to interact with the Terminal!

Reviewer: ```suggestion
When using the [Terminal](https://glosario.carpentries.org/en/#console), this step is known as `git config` with the `--global` option. In the next chapter we are going to interact with the Terminal!
```

:::::::::::::::::::::::::

<!--
If you are interested on this, [take a look to this chapter of the happygitwithr ebook](https://happygitwithr.com/hello-git).
-->

[git-privacy]: https://help.github.com/articles/keeping-your-email-address-private/

:::::::::::::::::::::::::::::::::::::::: keypoints

- Use the `{usethis}` package to configure a user name, email address, and other preferences once per machine.

> - Use `usethis::use_git_config()` to configure git in Rstudio.

Reviewer: You are mostly using Git as a proper noun (i.e. capitalising the G), but not always, I think either is fine, but be consistent.

Author: Thanks, I skimmed the docu that was the only one that needed an edit. https://github.com/epiverse-trace/git-rstudio-basics/pull/13/commits/bd5d2f0b72f1f882c549b1b9907c937e14b6a5ef

> - Use `usethis::git_default_branch_configure()` to define default branch name.

Reviewer: ```suggestion
- Use `usethis::git_default_branch_configure()` to define the default branch name.
```

::::::::::::::::::::::::::::::::::::::::::::::::::

-->

::::::::::::::::: instructor

When using `usethis::git_sitrep()`, check if there is no `✖ ...` line in the output with an error message. 

Some common errors are solved here:

- [Happy Git with R ebook](https://happygitwithr.com/https-pat.html#pat-troubleshooting)
- [Workbench documentation](https://carpentries.github.io/sandpaper-docs/github-pat.html#common-token-errors)

:::::::::::::::::::::::::

::::::::::::::::: callout

When using the [Terminal](https://glosario.carpentries.org/en/#console), this step is known as `git config` with the `--global` option. In the next chapter we are going to interact with the Terminal!

:::::::::::::::::::::::::

<!--
If you are interested on this, [take a look to this chapter of the happygitwithr ebook](https://happygitwithr.com/hello-git).
-->

[git-privacy]: https://help.github.com/articles/keeping-your-email-address-private/

:::::::::::::::::::::::::::::::::::::::: keypoints

- Use the `{usethis}` package to configure a user name, email address, and other preferences once per machine.

> - Use `usethis::use_git_config()` to configure git in Rstudio.

Author: ```suggestion
- Use `usethis::use_git_config()` to configure Git in Rstudio.
```
